### PR TITLE
[PropertyAccess] Use ArrayAdapter in debug mode

### PR DESCRIPTION
**Pull request by chalasr**
| Q             | A
| ------------- | ---
| Branch?       | 3.2
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #21894 
| License       | MIT
| Doc PR        | n/a

https://github.com/symfony/symfony/pull/21936/files?w=1

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 1]
```diff
@@ -1254,16 +1255,24 @@ private function registerCacheConfiguration(array $config, ContainerBuilder $con
         }
 
         if (method_exists(PropertyAccessor::class, 'createCache')) {
-            $propertyAccessDefinition = $container->register('cache.property_access', AdapterInterface::class);
+            $propertyAccessDefinition = $container->register('cache.property_access');
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: keep this line

**Review by nicolas-grekas — COMMENTED** [review]
:+1: ping @dunglas

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 2]
```diff
@@ -1256,14 +1257,21 @@ private function registerCacheConfiguration(array $config, ContainerBuilder $con
...
-            $propertyAccessDefinition->setArguments(array(null, null, $version, new Reference('logger', ContainerInterface::IGNORE_ON_INVALID_REFERENCE)));
             $propertyAccessDefinition->addTag('cache.pool', array('clearer' => 'cache.default_clearer'));
             $propertyAccessDefinition->addTag('monolog.logger', array('channel' => 'cache'));
+
+            if (!$container->getParameter('kernel.debug')) {
+                $propertyAccessDefinition->setFactory(array(PropertyAccessor::class, 'createCache'));
+                $propertyAccessDefinition->setArguments(array(null, null, $version, new Reference('logger', ContainerInterface::IGNORE_ON_INVALID_REFERENCE)));
+            } else {
+                $propertyAccessDefinition->setClass(ArrayAdapter::class);
+                $propertyAccessDefinition->setArguments(array(0));
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: why this line?

Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: in fact, can't you add false as second arg? (to store unserialized)

Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: in fact, can't we add `false` as second arg? (to store unserialized)

Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Removing this line gives:

> Symfony\Component\DependencyInjection\Exception\OutOfBoundsException: Cannot replace arguments if none have been configured yet.

>symfony/symfony/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolPass.php:74

`CachePoolPass.php:74` tries to replace arguments that are not set yet, for ArrayAdapter at least. Should it be fixed?

Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: `false` added as 2nd arg, works as well

Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: In fact, the ArrayAdapter should not have the cache.pool tag!

Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: indeed, fixed

**Review by nicolas-grekas — APPROVED** [review]
:+1:

**Review by xabbuh — APPROVED** [review]
👍

**Comment by nicolas-grekas** [comment]
Thank you @chalasr.

**Comment by teohhanhui** [comment]
Was there any profiling done to assess the performance impact of this change? (As opposed to stale checking.)

**Comment by nicolas-grekas** [comment]
Are you experiencing any issue? Please open new ticket, comments on closed can't be tracked.

**Pull request merged**